### PR TITLE
Setup schema.graphql, mapping(handler), schema, subgraph.yml file for ERC20Gateway contract

### LINKED ISSUE #275: Setup schema.graphql, mapping(handler), schema, subgraph.yml file for ERC20Gateway contract
- Define the entities in schema.graphq
```
DeclaredDepositIntent
      - id
      - token address
      - amount
      - nonce
      - beneficiary
      - fee gas price
      - fee gas limit
      - depositor
      - block number
      - block hash
      - contract address
      - updated time stamp (uts)

ConfirmedWithdrawIntent
      - id
      - message hash
      - block number
      - block hash
      - contract address
      - updated time stamp (uts)

ProvenGateway
      - id
      - proven block number
      - storage root
      - gateway address 
      - block number
      - block hash
      - contract address
      - updated time stamp (uts)
```
- Define handlers 
  - DeclaredDepositIntent, 
  - ConfirmedWithdrawIntent
  - ProvenGateway
- Add ABIs of
  - ERC20Gateway contract
- Events in subgraph.yaml
  - DepositIntentDeclared event
  - WithdrawIntentConfirmed event
  - GatewayProven event

### PR DESCRIPTION
**Pull request by Vindeep07**
Fixes #275 

- schema.graphql is populated with ERC20Gateway entities
- subgraph.yaml added with ERC20Gateway contract details
- ERC20GatewayMapping.ts
- generated folder to be compiled and deployed

### CODE REVIEW

File: subgraph/origin/subgraph.yaml [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+    source:
+      address: "0xdbbb5c822ccfac3ea49e32215565c6aafb257a30"
+      abi: ERC20Gateway
+    mapping:
+      kind: ethereum/events
+      apiVersion: 0.0.3
+      language: wasm/assemblyscript
+      entities:
+        - DepositIntentDeclared
+        - GatewayProven
```
Comment (abhayks1) [inline subgraph/origin/subgraph.yaml]: entity names are incorrect. It should align with schema.graphql entity names

File: subgraph/origin/subgraph.yaml [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+        - WithdrawIntentConfirmed
+      abis:
+        - name: ERC20Gateway
+          file: ../abi/ERC20Gateway.json
```
Comment (abhayks1) [inline subgraph/origin/subgraph.yaml:41]: Is this updated abi from mosaic-1 repo?

File: subgraph/origin/schema.graphql [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+  blockNumber: BigInt! # uint256
+  blockHash: Bytes! # bytes32
+  contractAddress: Bytes! # address
+  uts: BigInt! # uint256
+}
+
+type ProvenGateway @entity {
+  id: ID!
+  provenBlockNumber: BigInt! # uint256
+  gatewayAddress: Bytes! # address
```
Comment (abhayks1) [inline subgraph/origin/schema.graphql]:
```suggestion
  remoteGatewayAddress: Bytes! # address
```

Comment (abhayks1) [reply subgraph/origin/schema.graphql]: I think we can align ordering same as contract event.
provenBlockNumber can come after remoteGatewayAddress.
What do you think?

File: subgraph/origin/schema.graphql [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+  id: ID!
+  tokenAddress: Bytes! # address
+  amount: BigInt! # uint256
+  nonce: BigInt!  # uint256
+  beneficiary: Bytes! # address
+  feeGasPrice: BigInt! # uint256
+  feeGasLimit: BigInt! # uint256
+  depositor: Bytes! # address
+  blockNumber: BigInt! # uint256
+  blockHash: Bytes! # bytes32
```
Comment (abhayks1) [inline subgraph/origin/schema.graphql:21]: messageHash is missing here

File: subgraph/origin/generated/schema.ts [hunk 5]
```diff
@@ -0,0 +1,316 @@
+// THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
```
Comment (abhayks1) [inline subgraph/origin/generated/schema.ts]: schema.ts needs to be generated again after merge of PR #326

File: subgraph/origin/src/ERC20GatewayMapping.ts [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+  entity.contractAddress = event.address;
+  entity.uts = event.block.timestamp;
+
+  entity.save();
+}
+
+export function handleGatewayProven(
+  event: GatewayProven,
+): void {
+  // eslint-disable-next-line prefer-const
```
Comment (abhayks1) [inline subgraph/origin/src/ERC20GatewayMapping.ts]: You don't need to duplicate this. You can take it at the top of file only once.

**Review by abhayks1 — CHANGES_REQUESTED** [review]
Good work 👍 
There are inline comments.

File: subgraph/origin/src/ERC20GatewayMapping.ts (hunk 6 above)
Comment (Vindeep07) [reply subgraph/origin/src/ERC20GatewayMapping.ts]: Done.

File: subgraph/origin/subgraph.yaml (hunk 1 above)
Comment (Vindeep07) [reply subgraph/origin/subgraph.yaml]: Fixed.

File: subgraph/origin/schema.graphql (hunk 3 above)
Comment (Vindeep07) [reply subgraph/origin/schema.graphql]: Yah, we can align ordering same as contract. Changed.

File: subgraph/origin/schema.graphql (hunk 4 above)
Comment (Vindeep07) [reply subgraph/origin/schema.graphql:21]: `messageHash` added in `DeclaredDepositIntent` entity.

File: subgraph/origin/generated/schema.ts (hunk 5 above)
Comment (Vindeep07) [reply subgraph/origin/generated/schema.ts]: Yes. Understood.

File: subgraph/origin/subgraph.yaml (hunk 2 above)
Comment (Vindeep07) [reply subgraph/origin/subgraph.yaml:41]: Yes, its updated abi from mosaic-1.

File: subgraph/origin/schema.graphql [hunk 7]
```diff
@@ -6,3 +6,38 @@ type AvailableStateRoot @entity {
   contractAddress: Bytes! # address
   uts: BigInt! # uint256
 }
+
+type DeclaredDepositIntent @entity {
+  id: ID!
+  tokenAddress: Bytes! # address
```
Comment (abhayks1) [inline subgraph/origin/schema.graphql]: Please align the ordering as per contract event DepositIntentDeclared

File: subgraph/origin/schema.graphql [hunk 8]
```diff
@@ -6,3 +6,38 @@ type AvailableStateRoot @entity {
...
+  messageHash: Bytes! # bytes32
+  blockNumber: BigInt! # uint256
+  blockHash: Bytes! # bytes32
+  contractAddress: Bytes! # address
+  uts: BigInt! # uint256
+}
+
+type ProvenGateway @entity {
+  id: ID!
+  remoteGatewayAddress: Bytes! # address
```
Comment (abhayks1) [inline subgraph/origin/schema.graphql]:
```suggestion
  remoteGateway: Bytes! # address
```

File: subgraph/origin/subgraph.yaml [hunk 9]
```diff
@@ -21,3 +21,30 @@ dataSources:
         - event: StateRootAvailable(uint256,bytes32)
           handler: handleStateRootAvailable
       file: ./src/AnchorMapping.ts
+      
+  - kind: ethereum/contract
+    name: ERC20Gateway
+    network: dev
+    source:
+      address: "0xdbbb5c822ccfac3ea49e32215565c6aafb257a30"
```
Comment (abhayks1) [inline subgraph/origin/subgraph.yaml]: Update in ticket that contract address needs to be updated after deployment and tag @deepesh-kn

File: subgraph/origin/src/ERC20GatewayMapping.ts [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+/* eslint-disable prefer-const */
+/* eslint-disable prefer-template */
+export function handleDepositIntentDeclared(
+  event: DepositIntentDeclared,
+): void {
+  let entity = new DeclaredDepositIntent(
+    event.transaction.hash.toHex() + '_' + event.logIndex.toString(),
+  );
+
+  entity.tokenAddress = event.params.valueToken;
```
Comment (abhayks1) [inline subgraph/origin/src/ERC20GatewayMapping.ts]: this can be renamed to valueToken

File: subgraph/origin/package-lock.json [hunk 11]
```diff
@@ -184,7 +184,8 @@
         "safe-buffer": {
```
Comment (abhayks1) [inline subgraph/origin/package-lock.json:184]: Update the package.json with the correct subgraph name.
```
mosaic/{origin|aux}-erc20gateway 
```
You need to run `npm i` again after updating name.

File: subgraph/origin/src/ERC20GatewayMapping.ts [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+  WithdrawIntentConfirmed,
+} from '../generated/ERC20Gateway/ERC20Gateway';
+import {
+  DeclaredDepositIntent,
+  ConfirmedWithdrawIntent,
+  ProvenGateway,
+} from '../generated/ERC20GatewaySchema';
+
+/* eslint-disable prefer-const */
+/* eslint-disable prefer-template */
```
Comment (abhayks1) [inline subgraph/origin/src/ERC20GatewayMapping.ts]: This can be combined in 1 line.

**Review by abhayks1 — CHANGES_REQUESTED** [review]
There are few final comments.

File: subgraph/origin/schema.graphql (hunk 7 above)
Comment (Vindeep07) [reply subgraph/origin/schema.graphql]: Entities have been aligned as per contract events.

File: subgraph/origin/schema.graphql (hunk 8 above)
Comment (Vindeep07) [reply subgraph/origin/schema.graphql]: Fixed.

File: subgraph/origin/subgraph.yaml (hunk 9 above)
Comment (Vindeep07) [reply subgraph/origin/subgraph.yaml]: Done.

File: subgraph/origin/src/ERC20GatewayMapping.ts (hunk 10 above)
Comment (Vindeep07) [reply subgraph/origin/src/ERC20GatewayMapping.ts]: Changed.

File: subgraph/origin/package-lock.json (hunk 11 above)
Comment (Vindeep07) [reply subgraph/origin/package-lock.json:184]: Done.

File: subgraph/origin/src/ERC20GatewayMapping.ts (hunk 12 above)
Comment (Vindeep07) [reply subgraph/origin/src/ERC20GatewayMapping.ts]: Fixed.

**Review by abhayks1 — APPROVED** [review]
LGTM 👍

**Pull request merged**